server/src/main.rs

```diff
@@ -0,0 +1,201 @@
+                    TextDocumentSyncOptions {
+                        open_close: Some(true),
+                        change: Some(TextDocumentSyncKind::FULL), // or TextDocumentSyncKind::INCREMENTAL
+                        save: Some(TextDocumentSyncSaveOptions::SaveOptions(SaveOptions {
+                            include_text: Some(true),
+                        })),
+                        ..Default::default()
+                    },
+                )),
+                definition_provider: Some(OneOf::Left(true)),
+                hover_provider: Some(HoverProviderCapability::Simple(true)),
+                ..ServerCapabilities::default()
+            },
+        })
+    }
+
+    async fn initialized(&self, _: InitializedParams) {
+        // Compile all files to get the initial compilation state
+        let (initial_state, options) = self.compile_slice_files().await;
+        let mut compilation_state_guard = self.compilation_state.lock().await;
+        *compilation_state_guard = initial_state;
+        let mut compilation_options_guard = self.compilation_options.lock().await;
+        *compilation_options_guard = options;
+
+        self.client
+            .log_message(MessageType::INFO, "Slice language server initialized!")
+            .await;
+    }
+
+    async fn shutdown(&self) -> tower_lsp::jsonrpc::Result<()> {
+        Ok(())
+    }
+
+    async fn goto_definition(
+        &self,
+        params: GotoDefinitionParams,
+    ) -> tower_lsp::jsonrpc::Result<Option<GotoDefinitionResponse>> {
+        let uri = params.text_document_position_params.text_document.uri;
+        let position = params.text_document_position_params.position;
+        let state = self.compilation_state.lock().await;
+
+        if let Some(location) = get_definition_span(&state, uri, position) {
+            let start = Position {
+                line: (location.start.row - 1) as u32,
+                character: (location.start.col - 1) as u32,
+            };
+
+            let end = Position {
+                line: (location.end.row - 1) as u32,
+                character: (location.end.col - 1) as u32,
+            };
+
+            Ok(Some(GotoDefinitionResponse::Scalar(Location {
+                uri: Url::from_file_path(location.file).unwrap(),
+                range: Range::new(start, end),
+            })))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn hover(&self, params: HoverParams) -> tower_lsp::jsonrpc::Result<Option<Hover>> {
+        let uri = params.text_document_position_params.text_document.uri;
+        let position = params.text_document_position_params.position;
+        let state = self.compilation_state.lock().await;
+        if let Some(info) = get_hover_info(&state, uri, position) {
+            Ok(Some(Hover {
+                contents: HoverContents::Scalar(MarkedString::String(info)),
+                range: None,
+            }))
+        } else {
+            Ok(None)
+        }
```

Comment:
Since both branches always return `Ok`, and the `None` case also returns a `None`, you could refactor this as:
```suggestion
        Ok(get_hover_info(&state, uri, position).map(|info| {
            Hover {
                contents: HoverContents::Scalar(MarkedString::String(info)),
                range: None,
            }
        }))
```